src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp

```diff
@@ -247,109 +249,352 @@ void  ShenandoahGeneration::prepare_regions_and_collection_set(bool concurrent)
+      //
+      //  3. Next, we calculate a young-gen evacuation budget, which is the smaller of the two quantities mentioned
+      //     above.  old_gen_memory_available_to_be_loaned is calculated as:
+      //       old_gen->available - (advance-promotion-reserve + old-gen_evacuation_reserve)
+      //
+      //  4. When choosing the collection set, special care is taken to assure that the amount of loaned memory required to
+      //     hold the results of evacuation is smaller than the total memory occupied by the regions added to the collection
+      //     set.  We need to take these precautions because we do not know how much memory will be reclaimed by evacuation
+      //     until after the collection set has been constructed.  The algorithm is as follows:
+      //
+      //     a. We feed into the algorithm (i) young available at the start of evacuation and (ii) the amount of memory
+      //        loaned from old-gen that is available to hold the results of evacuation.
+      //     b. As candidate regions are added into the young-gen collection set, we maintain accumulations of the amount
+      //        of memory spanned by the collection set regions and the amount of memory that must be reserved to hold
+      //        evacuation results (by multiplying live-data size by ShenandoahEvacWaste).  We process candidate regions
+      //        in order of decreasing amounts of garbage.  We skip over (and do not include into the collection set) any
+      //        regions that do not satisfy all of the following conditions:
+      //
+      //          i. The amount of live data within the region as scaled by ShenandoahEvacWaste must fit within the
+      //             relevant evacuation reserve (live data of old-gen regions must fit within the old-evac-reserve, live
+      //             data of young-gen tenure-aged regions must fit within the advance promotion reserve, live data within
+      //             other young-gen regions must fit within the youn-gen evacuation reserve).
+      //         ii. The accumulation of memory consumed by evacuation must not exceed the accumulation of memory reclaimed
+      //             through evacuation by more than young-gen available.
+      //        iii. Other conditions may be enforced as appropriate for specific heuristics.
+      //
+      //       Note that regions are considered for inclusion in the selection set in order of decreasing amounts of garbage.
+      //       It is possible that a region with a larger amount of garbage will be rejected because it also has a larger
+      //       amount of live data and some region that follows this region in candidate order is included in the collection
+      //       set (because it has less live data and thus can fit within the evacuation limits even though it has less
+      //       garbage).
+
+      size_t young_evacuation_reserve = (young_generation->soft_max_capacity() * ShenandoahEvacReserve) / 100;
```

Comment:
Should we try to align this more closely with how `ShenandoahFreeSet` behaves? It uses `max_capacity`, for example. It could also tell us exactly how much it really reserved.

Reply:
I will change this.